Commands/Services/MailingService.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System.Net;
+using System.Net.Mail;
+
+using Commands.Configuration;
+
+using Domain;
+
+using Microsoft.Extensions.Options;
+
+namespace Commands.Services;
+
+public interface IMailingService
+{
+    Task<IEnumerable<SendError>> SendMailAsync(IEnumerable<Member> members, Message message, CancellationToken ct);
+}
+
+public class MailingService : IMailingService
+{
+    private readonly MailingConfiguration _config;
+    public MailingService(IOptions<MailingConfiguration> options)
+    {
+        if (options is null)
+            throw new ArgumentNullException(nameof(options));
+
+        _config = options.Value;
+    }
+
+    public async Task<IEnumerable<SendError>> SendMailAsync(IEnumerable<Member> members, Message message, CancellationToken ct)
+    {
+        var sender = new MailAddress(_config.Sender.Email, _config.Sender.Name);
+        var recipients = members.Select(m => new MailAddress(m.Email ?? throw new Exception("Can't send email to members without email address"), m.Name));
+
+        var messages = recipients.Select(r => new MailMessage(sender, r)
+        {
+            Subject = message.Subject,
+            Body = message.Body
+
+        });
+
+        var errors = new List<SendError>();
+
+        var mailTasks = messages.Select(async msg =>
+        {
+            try
+            {
+                using var client = GetSmtpClient;
```

Comment:
Why would you create the smtpclient for every message? The configuration doesn't change, during sending the e-mail, does it?

Reply:
Because SmtpClient is not thread safe, see https://learn.microsoft.com/en-us/dotnet/api/system.net.mail.smtpclient?view=net-7.0&redirectedfrom=MSDN#remarks
Since we start a ln async operation here for every email recipient, this will likely cause problems